src/mumble/Zeroconf.cpp

```diff
@@ -0,0 +1,302 @@
+// Copyright 2005-2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#include "Zeroconf.h"
+
+#define SYMBOL_EXISTS(symbol) (GetProcAddress(handle, symbol))
+
+Zeroconf::Zeroconf() : m_ok(false) {
+	qRegisterMetaType< uint16_t >("uint16_t");
+	qRegisterMetaType< BonjourRecord >("BonjourRecord");
+	qRegisterMetaType< QList< BonjourRecord > >("QList<BonjourRecord>");
+#ifdef Q_OS_WIN64
+	static bool winDnsUnavailable = false;
+	if (!winDnsUnavailable) {
+		auto handle = GetModuleHandle(L"dnsapi.dll");
+		if (handle) {
+			if (SYMBOL_EXISTS("DnsServiceBrowse") && SYMBOL_EXISTS("DnsServiceBrowseCancel")
+				&& SYMBOL_EXISTS("DnsServiceResolve") && SYMBOL_EXISTS("DnsServiceResolveCancel")
+				&& SYMBOL_EXISTS("DnsServiceFreeInstance")) {
+				m_ok = true;
+				return;
+			}
+		}
+
+		winDnsUnavailable = true;
+		qWarning("Zeroconf: Native mDNS/DNS-SD API not available, falling back to third-party API");
+	}
+
+	static bool dnssdLoadFailed = false;
+	if (!dnssdLoadFailed) {
+		auto handle = LoadLibrary(L"dnssd.DLL");
+		if (!handle) {
+			dnssdLoadFailed = true;
+			qWarning("Zeroconf: Failed to load dnssd.dll, assuming third-party API is not available");
+			return;
+		}
+		FreeLibrary(handle);
+	}
+#endif
```

Comment:
What happens if this code is not present? Will it fallback to Bonjour?

Reply:
Correct, the last lines of the constructor (outside the `#ifdef` block) are always executed.